packages/react-pages/src/node/dynamic-modules/PageStrategy.ts

```diff
@@ -0,0 +1,297 @@
+        if (page.data[key])
+          throw Error(
+            `addPageData conflict: data with key "${key}" already exists`
+          )
+        page.data[key] = dataPath
+      }
+
+      if (staticData) {
+        if (page.staticData[key])
+          throw Error(
+            `addPageData conflict: staticData with key "${key}" already exists`
+          )
+        page.staticData[key] = staticData
+      }
+
+      if (currentFile) {
+        let dataKeys = currentFile.dataKeys.get(pageId)
+        if (!dataKeys) currentFile.dataKeys.set(pageId, (dataKeys = []))
+        dataKeys.push(key)
+      }
+
+      if (existingPage) {
+        changedPages.add(pageId)
+        emitChange()
+      }
+    }
+
+    function removePageData(dataKeys: string[], pageId: string) {
+      const page = pageCache[pageId]
+
+      dataKeys.forEach((key) => {
+        delete page.data[key]
+        delete page.staticData[key]
+      })
+
+      if (hasKeys(page.data) || hasKeys(page.staticData)) {
+        changedPages.add(pageId)
+        emitChange()
+      } else {
+        delete pageCache[pageId]
+      }
+    }
+
+    function hasKeys(obj: object) {
+      for (const _ in obj) return true
+      return false
+    }
+
+    function watchFiles(
+      baseDir: string,
+      arg2?: string | string[] | FileHandler,
+      arg3?: FileHandler
+    ) {
+      let globs: string[]
+      let handler: FileHandler
+
+      if (typeof arg2 === 'function') {
+        globs = ['**/*']
+        handler = arg2
+      } else {
+        globs = Array.isArray(arg2) ? arg2 : [arg2 || '**/*']
+        handler = arg3 || defaultPageLoader
+      }
+
+      // Strip trailing slash and make absolute
+      baseDir = path.resolve(pagesDir, baseDir)
+
+      const processFile = async (
+        type: 'add' | 'change' | 'unlink',
+        file: File
+      ) => {
+        file.dataKeys.forEach(removePageData)
+        if (type !== 'unlink') {
+          if (type === 'change') {
+            file.content = null
+            file.dataKeys.clear()
+          }
+          currentFile = file
+          try {
+            const pageData = await handler(file)
+            if (pageData) {
+              addPageData(pageData)
+            }
+          } finally {
+            currentFile = null
+          }
+        }
+      }
+
+      watchers.add(
+        chokidar
+          .watch(globs, {
+            cwd: baseDir,
+            ignored: ['**/node_modules/**/*'],
+          })
+          .on('all', (type, filePath) => {
+            if (type !== 'addDir' && type !== 'unlinkDir') {
+              filePath = path.join(baseDir, filePath)
+
+              const file =
+                fileCache[filePath] ||
+                (fileCache[filePath] = new File(filePath, baseDir))
+
+              if (type === 'unlink') {
+                delete fileCache[file.path]
+              } else if (file.queued) {
+                return
+              }
+
+              file.queued = true
```

Comment:
Could you make the code clearer by not queuing unlinked files. That way the `processFile` can also be simplifed.

Reply:
You can move the `file.dataKeys.forEach(removePageData)` in this function. So that all the login of unlinked files processing is here.

Reply:
I don't think so. We need to wait for `add`/`change` events to be processed before the `unlink` event can be. Otherwise, we'll end up with deleted pages still being cached.